binderhub/repoproviders.py

```diff
@@ -83,6 +91,141 @@ def get_build_slug(self):
+        # escape the name and replace dashes with something else.
+        return '-'.join(p.replace('-', '_-') for p in parts)
+
+
+class GitLabRepoProvider(RepoProvider):
+    """GitLab provider.
+
+    GitLab allows nested namespaces (eg. root/project/component/repo) thus we need to urlescape the namespace of this
+    repo.  Users must provide a spec that matches the following form.
+
+    <url-escaped-namespace>/<unresolved_ref>
+
+    eg:
+    group%2Fproject%2Frepo/master
+    """
+
+    name = Unicode('GitLab')
+    hostname = Unicode('gitlab.com')
+
+    access_token = Unicode(config=True,
+        help="""GitLab OAuth2 access token for authentication with the GitLab API
+
+        For use with client_secret.
+        Loaded from GITLAB_ACCESS_TOKEN env by default.
+        """
+    )
+    @default('access_token')
+    def _access_token_default(self):
+        return os.getenv('GITLAB_ACCESS_TOKEN', '')
+
+    private_token = Unicode(config=True,
+        help="""GitLab private token for authentication with the GitLab API
+
+        Loaded from GITLAB_PRIVATE_TOKEN env by default.
+        """
+    )
+    @default('private_token')
+    def _private_token_default(self):
+        return os.getenv('GITLAB_PRIVATE_TOKEN', '')
+
+    auth = Dict(
+        help="""Auth parameters for the GitLab API access
+
+        Populated from access_token, private_token
+    """
+    )
+    @default('auth')
+    def _default_auth(self):
+        auth = {}
+        for key in ('access_token', 'private_token'):
+            value = getattr(self, key)
+            if value:
+                auth[key] = value
+        return auth
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        quoted_namespace, unresolved_ref = self.spec.split('/', 1)
+        self.namespace = urllib.parse.unquote(quoted_namespace)
+        self.unresolved_ref = urllib.parse.unquote(unresolved_ref)
+        if not self.unresolved_ref:
+            raise ValueError("An unresolved ref is required")
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        if hasattr(self, 'resolved_ref'):
+            return self.resolved_ref
+
+        namespace = urllib.parse.quote(self.namespace, safe='')
+        client = AsyncHTTPClient()
+        api_url = "https://{hostname}/api/v4/projects/{namespace}/repository/commits/{ref}".format(
+            namespace=namespace, ref=urllib.parse.quote(self.unresolved_ref, safe=''), hostname=self.hostname
+        )
+        self.log.debug("Fetching %s", api_url)
+
+        if self.auth:
+            # Add auth params. After logging!
+            api_url = url_concat(api_url, self.auth)
+
+        try:
+            resp = yield client.fetch(api_url, user_agent="BinderHub")
+        except HTTPError as e:
+            if e.code == 404:
+                return None
+            else:
+                raise
+
+        ref_info = json.loads(resp.body.decode('utf-8'))
+        self.resolved_ref = ref_info['id']
```

Comment:
Will the ref_info always have 'id'? What if the given branch or commit / tag does not exist? Will gitlab return 404 for 'em?

Reply:
Yeah Gitlab 404s